lib/evmone_precompiles/kzg.cpp

```diff
@@ -0,0 +1,155 @@
+            0x64044f659f0ee1e9, 0x0ed54f48d5a1caa7}}},
+    {{ONE, {}}}};
+
+/// The point from the G2 series, index 1 of the Ethereum KZG trusted setup,
+/// i.e. [s]₂ where s is the trusted setup's secret.
+/// Affine coordinates in Montgomery form.
+constexpr blst_p2_affine KZG_SETUP_G2_1{
+    {{{0x6120a2099b0379f9, 0xa2df815cb8210e4e, 0xcb57be5577bd3d4f, 0x62da0ea89a0c93f8,
+          0x02e0ee16968e150d, 0x171f09aea833acd5},
+        {0x11a3670749dfd455, 0x04991d7b3abffadc, 0x85446a8e14437f41, 0x27174e7b4e76e3f2,
+            0x7bfa6dd397f60a20, 0x02fcc329ac07080f}}},
+    {{{0xaa130838793b2317, 0xe236dd220f891637, 0x6502782925760980, 0xd05c25f60557ec89,
+          0x6095767a44064474, 0x185693917080d405},
+        {0x549f9e175b03dc0a, 0x32c0c95a77106cfe, 0x64a74eae5705d080, 0x53deeaf56659ed9e,
+            0x09a1d368508afb93, 0x12cf3a4525b5e9bd}}}};
+
+/// Load and validate an element from the group order field.
+std::optional<blst_scalar> validate_scalar(std::span<const std::byte, 32> b) noexcept
+{
+    blst_scalar v;
+    blst_scalar_from_bendian(&v, reinterpret_cast<const uint8_t*>(b.data()));
+    return blst_scalar_fr_check(&v) ? std::optional{v} : std::nullopt;
+}
+
+/// Uncompress and validate a point from G1 subgroup.
+std::optional<blst_p1_affine> validate_G1(std::span<const std::byte, 48> b) noexcept
+{
+    blst_p1_affine r;
+    if (blst_p1_uncompress(&r, reinterpret_cast<const uint8_t*>(b.data())) != BLST_SUCCESS)
+        return std::nullopt;
+    if (!blst_p1_affine_in_g1(&r))  // Subgroup check is required by the spec but not testable.
+        return std::nullopt;
+    return r;
+}
+
+blst_p1_affine add_or_double(const blst_p1_affine& p, const blst_p1& q) noexcept
+{
+    blst_p1 r;
+    blst_p1_add_or_double_affine(&r, &q, &p);
+    blst_p1_affine ra;
+    blst_p1_to_affine(&ra, &r);
+    return ra;
+}
+
+blst_p1 mult(const blst_p1& p, const blst_scalar& v) noexcept
+{
+    blst_p1 r;
+    blst_p1_mult(&r, &p, v.b, 255);
+    return r;
+}
+
+blst_p2_affine add(const blst_p2_affine& p, const blst_p2& q) noexcept
+{
+    blst_p2 r;
+    blst_p2_add_affine(&r, &q, &p);
+    blst_p2_affine ra;
+    blst_p2_to_affine(&ra, &r);
+    return ra;
+}
+
+blst_p2 mult(const blst_p2& p, const blst_scalar& v) noexcept
+{
+    blst_p2 r;
+    blst_p2_mult(&r, &p, v.b, 255);
+    return r;
+}
+
+bool pairings_verify(
+    const blst_p1_affine& a1, const blst_p1_affine& b1, const blst_p2_affine& b2) noexcept
+{
+    blst_fp12 left;
+    blst_aggregated_in_g1(&left, &a1);
+    blst_fp12 right;
+    blst_miller_loop(&right, &b2, &b1);
+    return blst_fp12_finalverify(&left, &right);
+}
+}  // namespace
+
+bool kzg_verify_proof(const std::byte versioned_hash[32], const std::byte z[32],
+    const std::byte y[32], const std::byte commitment[48], const std::byte proof[48]) noexcept
+{
+    std::byte computed_versioned_hash[32];
+    sha256(computed_versioned_hash, commitment, 48);
+    computed_versioned_hash[0] = std::byte{0x01};
```

Comment:
Note to self: 0x01 is `VERSIONED_HASH_VERSION_KZG`

Reply:
The constant name in use now.